### DISCUSSION TITLE
New `nongeneric` keyword to prevent repeated initialization of statics fields in generic types

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#5587] [Language Ideas]
Consider this:
```cs
class Bad<T>
{
  static readonly Regex _rgx = new("...");
}
var b1 = new Bad<int>(); // one Regex created - we're good
var b2 = new Bad<long>(); // ANOTHER Regex created - bad
// compare to:
class Good<T>
{
  nongeneric static readonly Regex _rgx = new("...");
}
var g1 = new Good<int>(); // one Regex created - we're good
var g2 = new Good<long>(); // no more Regex created - we're still good
```
This new keyword will be placed right before `static`.
Implementation: 

- either create a special hidden non-generic class to house these static fields; 
- - this has the downside of "disconnecting" the class as it's written in code from the class as it's defined in IL, and is therefore undesirable;
- add a runtime flag of some sort to mark these fields as "non generic" and then runtime can take care of avoiding re-initializing them;
- restrict this to properties, and the compiler would create a hidden non-generic class to house the backing field:
- - that way, the property works as expected and is found in the same place both in the code and in the IL.

Downsides: none I can think of, other than added complexity.

**Comment by CyrusNajmabadi** [comment]
> Downsides: none I can think of, other than added complexity.

As mentioned in our prior conversation this would break our desire that top-level symbols map to their metadata counterparts.  This would not be the case here.

**Reply by TahirAhmadov** [reply]
It would, with the runtime flag you had suggested.

**Reply by CyrusNajmabadi** [reply]
i was addressing this guy:

> create a special hidden non-generic class to house these static fields; 

:)

**Reply by IS4Code** [reply]
If the only remaining option is via a runtime flag/attribute, perhaps this is more of a "runtime" proposal, similarly to how `[ThreadStatic]` modifies the semantics of such a field. Such a "`[NonGeneric]`" or "`[NonInstantiated]`" attribute on a field could simply mean that all generic instantiations use a shared storage for that field. The type initializer will still be executed multiple times however, so we'd need non-instantiated type initializers as well. This quickly become quite complicated and hard to use.

Why not have non-instantiated properties then instead?

```cs
static nongeneric Regex _rgx { get; } = new("...");
```
becoming
```cs
static Regex _rgx => Inner._rgx;

static class Inner // without any type parameters in CIL
{
    public static readonly Regex _rgx = new("...");
}
```
No issue with symbol mapping, a property remains a property.

**Reply by TahirAhmadov** [reply]
@IllidanS4 the problem is that the `Inner` class has to be outside of your generic class, which is hard to read and write. In the real world, this means nobody will do that - and performance suffers.
However, it's not just performance - sometimes you need static non-generic state. Again, you could create a separate class to house those, but that's unwieldly.
Another problem with the separate class workaround is that you lose the ability to make this stuff private; at best, you can make it internal, but that's not the same as a private static field.
Regarding type initializer, we can either create another initializer (`static nongeneric Class() { ... }`), or not do it and just let the existing static ctor do some kind of lazy initialization, or add a bool parameter which is true when it's the first time this generic class is initialized (`static Class(bool firstCall) { if(firstCall) _nonGenericField = ...; }`).

**Reply by IS4Code** [reply]
@TahirAhmadov `Inner` doesn't have to be outside the class, if we can convince the compiler not to add the outer class's type parameters to it. That could be a separate proposal (and I hope it is!).

Anyway I was really arguing in favour of the compiler doing this transformation. With a property, all these implementation details are hidden and do not clutter the code, while keeping the result stable in terms of metadata.

**Reply by TahirAhmadov** [reply]
Oh I see what you mean. That's actually a good idea - I'll update the OP. 
Regarding inner-outer, I don't think the compiler can decide that - it's how metadata is structured in .NET and would require a runtime change. You can try it yourself - create a small program with a "non-generic" nested class inside a generic outer class, look at it using reflection API, and you will see that there is no way to make a **real** non-generic nested class - it always "inherits" type context from outer class.

**Comment by bernd5** [comment]
As an easy workaround I would simply inherit the non generic stuff...

e.g.:

```csharp
class BaseXXX
{
  static readonly Regex _rgx = new("...");
}

class XXX<T> : BaseXXX
{
}
````

**Reply by TahirAhmadov** [reply]
I do that already in many cases, actually. Obviously this is only possible if your class inherits from `object`.
I was motivated to request this addition because of the discussion about static locals over at #832. To make those work in generic methods, we got into the detail of "is it possible to define non-generic static locals in a generic method", and that's what led me to create this proposal.

**Reply by mburbea** [reply]
You can also use an interface since C#8.
```cs
interface IRegexHolder
{
     protected static readonly Regex _rgx = new("...");
}

class XXX<T> : IRegexHolder
{
     public static void SomeMethodThatUsesTheInterfacesStaticField() => Console.WriteLine(IRegexHolder._rgx);
}
```

**Reply by TahirAhmadov** [reply]
@mburbea great suggestion, thank you - I hadn't thought of this before.
Still, I'm hopeful to get static locals and for those, we would need a keyword, because those will be defined inside a method body.

**Reply by mburbea** [reply]
It's pretty handy trick I've started to use in a few places where I need some static data that would be shared between all generic instances, but isn't necessarily used by all decedents of a base class.  

Code is read far more often than it's written so I question how useful that feature is. Yes, there are cases where I have a static that is really only used by one method, but declaring it in the method can raise the question of what does something like::
```
static int Foo;
if(Foo != 0){
}
```
Is that if block dead code? How can `Foo` not be 0? Is this an error because `Foo` does not have a value initialized?
